File: day16/day16.py
#!/usr/bin/env python3

import math
import sys


class Packet(object):
    def __init__(self, datastream):
        self._datastream = datastream
        self._ptr = 0

        self.version = self.next_int(3)
        self.typeid = self.next_int(3)

        self.subpackets = []

        if self.typeid == 4:
            payload = []
            while True:
                bits = self.next(5)
                payload.append(bits[1:])
                if bits[0] == '0':
                    break

            self.value = int(''.join(payload), 2)
        else:
            length_type = self.next()
            if length_type == '0':
                num_bits = self.next_int(15)
                payload = self.next(num_bits)

                while payload:
                    subpacket = Packet(payload)
                    payload = subpacket.remaining()
                    self.subpackets.append(subpacket)
            else:
                num_packets = self.next_int(11)
                payload = self.remaining()
                for i in range(num_packets):
                    subpacket = Packet(payload)
                    payload = subpacket.remaining()
                    self.subpackets.append(subpacket)

                self._datastream = payload
                self._ptr = 0

            if self.typeid == 0:
                self.value = sum(self.subpacket_values())
            elif self.typeid == 1:
                self.value = math.prod(self.subpacket_values())
            elif self.typeid == 2:
                self.value = min(self.subpacket_values())
            elif self.typeid == 3:
                self.value = max(self.subpacket_values())
            elif self.typeid == 5:
                self.value = int(self.subpackets[0].value
                                 > self.subpackets[1].value)
            elif self.typeid == 6:
                self.value = int(self.subpackets[0].value
                                 < self.subpackets[1].value)
            else:
                self.value = int(self.subpackets[0].value
                                 == self.subpackets[1].value)

    def subpacket_values(self):
        return (p.value for p in self.subpackets)

    def all_subpackets(self):
        subpackets = []
        for subpacket in self.subpackets:
            subpackets.append(subpacket)
            subpackets.extend(subpacket.all_subpackets())
        return subpackets

    def pprint(self, level=0):
        print(' ' * (4 * level) +
              f'Version={self.version}, Type={self.typeid}, '
              f'Value={self.value or "None"}')
        for subpacket in self.subpackets:
            subpacket.pprint(level + 1)

    def next(self, n=1):
        data = self._datastream[self._ptr:self._ptr + n]
        self._ptr += n
        return data

    def next_int(self, n=1):
        return int(self.next(n), 2)

    def remaining(self):
        return self._datastream[self._ptr:]

    def __repr__(self):
        return (f'Packet(version={self.version}, type={self.typeid}, '
                f'value={self.value or "None"}, '
                f'{len(self.subpackets)} subpacket(s)')


def main():
    with open(sys.argv[1]) as f:
        datastream = ''.join(f'{int(i, 16):04b}' for i in f.read().strip())

    packet = Packet(datastream)

    print('Sum of packet versions:', packet.version +
          sum(p.version for p in packet.all_subpackets()))
    print('Packet value:', packet.value)


if __name__ == '__main__':
    main()

File: day16/day16.rs
struct Packet {
    version: u8,
    version_sum: u32,
    typeid: u8,
    subpackets: Vec<Packet>,
    value: u64,
}

impl Packet {
    pub fn from_bitstream(bits: &mut String) -> Self {
        let version = u8::from_str_radix(bits.drain(..3).as_str(), 2).unwrap();
        let typeid = u8::from_str_radix(bits.drain(..3).as_str(), 2).unwrap();

        let mut subpackets = Vec::new();
        let value: u64 = if typeid == 4 {
            let mut literal = String::new();
            loop {
                let cont = bits.drain(..1).next().unwrap();
                literal.push_str(bits.drain(..4).as_str());
                if cont == '0' {
                    break;
                }
            }

            u64::from_str_radix(&literal, 2).unwrap()
        } else {
            let length_type = bits.drain(..1).next().unwrap();
            if length_type == '0' {
                let num_bits = usize::from_str_radix(bits.drain(..15).as_str(), 2).unwrap();
                let mut remaining: String = bits.drain(..num_bits).collect();
                while !remaining.is_empty() {
                    let subpacket = Packet::from_bitstream(&mut remaining);
                    subpackets.push(subpacket);
                }
            } else {
                let num_packets = usize::from_str_radix(bits.drain(..11).as_str(), 2).unwrap();
                for _ in 0..num_packets {
                    let subpacket = Packet::from_bitstream(bits);
                    subpackets.push(subpacket);
                }
            }

            let subpacket_values = subpackets.iter().map(|p| p.value);
            match typeid {
                0 => subpacket_values.sum::<u64>(),
                1 => subpacket_values.product::<u64>(),
                2 => subpacket_values.min().unwrap(),
                3 => subpacket_values.max().unwrap(),
                _ => {
                    let first = subpackets[0].value;
                    let second = subpackets[1].value;
                    match typeid {
                        5 => (first > second) as u64,
                        6 => (first < second) as u64,
                        _ => (first == second) as u64,
                    }
                }
            }
        };

        let version_sum: u32 =
            version as u32 + subpackets.iter().map(|p| p.version_sum as u32).sum::<u32>();

        Self {
            version,
            version_sum,
            typeid,
            subpackets,
            value,
        }
    }

    pub fn from_hexstream(hexstream: &str) -> Self {
        let mut bitstream = String::new();
        hexstream
            .chars()
            .map(|h| format!("{:04b}", u8::from_str_radix(&h.to_string(), 16).unwrap()))
            .for_each(|bin| bitstream.push_str(&bin));

        Self::from_bitstream(&mut bitstream)
    }

    #[allow(dead_code)]
    fn pprint_level(&self, level: usize) {
        println!(
            "{}Version={}, type={}, value={}",
            if level > 0 {
                " ".repeat(level * 4)
            } else {
                "".to_string()
            },
            self.version,
            self.typeid,
            self.value
        );
        for packet in &self.subpackets {
            packet.pprint_level(level + 1);
        }
    }

    #[allow(dead_code)]
    pub fn pprint(&self) {
        self.pprint_level(0);
    }
}

impl std::fmt::Debug for Packet {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(
            f,
            "Packet(version={}, typeid={}, value={:?} subpackets={:?})",
            self.version, self.typeid, self.value, self.subpackets
        )
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let packet = Packet::from_hexstream(std::fs::read_to_string(&args[1]).unwrap().trim());

    println!("Sum of packet versions: {}", packet.version_sum);
    println!("Packet value: {}", packet.value);
}
